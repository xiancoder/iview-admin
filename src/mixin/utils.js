export default {
    methods: {
        go (name, obj) {
            if (obj) {
                this.$router.push({ name, obj });
            } else if (name) {
                this.$router.push({ name });
            } else {
                window.location.reload();
            }
        },
        saveParamState (obj) {
            const time = new Date().getTime()
            const name = this.$router.history.current.name
            window.localStorage.setItem(name + time, JSON.stringify(obj));
            const query = { 'search': time }
            // this.$router.replace({ name, query }) // 不能回退
            this.$router.push({ name, query })
        },
        getParamState () {
            const time = this.$route.query.search
            const name = this.$router.history.current.name
            if (time) {
                try {
                    const obj = window.localStorage.getItem(name + time);
                    return JSON.parse(obj)
                } catch (e) {
                    return {}
                }
            }
            if (window.history.length <= 1) {
                window.localStorage.clear()
            }
            return {}
        },
        alert (msg) {
            this.$Message.info({
                content: msg || '处理中，请稍后'
            });
        },
        success (msg) {
            this.$Message.info({
                content: msg || '保存成功'
            });
        },
        error (msg) {
            this.$Message.info({
                content: msg || '保存失败',
                duration: 10,
                closable: true
            });
        },
        confirm (msg) {
            return new Promise((resolve, reject) => {
                this.$Modal.confirm({
                    title: '请确认',
                    content: '<p>' + msg + '</p>',
                    onOk: resolve || function () {},
                    onCancel: reject || function () {}
                });
            });
        }
    }
};
