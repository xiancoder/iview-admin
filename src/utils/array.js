'use strict';

// =====================
// 数组去重唯一化
// =====================
// liuyp 2018年12月20日11:28:08
export const unique = (array) => {
    return Array.from(new Set(array))
}
// 数组去重唯一化(对象数组)
// =====================
// liuyp 2019年8月21日15:36:43
export const uniqueObj = (data, fun) => {
    var newData = [];
    fun = fun || function (a, b) { return a === b }
    for (var i = 0, len = data.length; i < len; i++) {
        var flag = 1;
        for (var j = 0, len2 = newData.length; j < len2; j++) {
            if (fun(newData[j], data[i])) { flag = 0; break; }
        }
        if (flag === 1) newData.push(data[i])
    }
    return newData;
}
// 数组去重唯一化 (完美版)(推荐)
// @param arr {array} 待处理数组
// @param [att] {string} 唯一的属性 可不填
// =====================
// liuyp 2019年9月20日11:15:33
export const arrayUnique = (arr, att) => {
    let resArr = []
    let json = {}
    for (let i = 0; i < arr.length; i++) {
        if (!att) {
            if (!json['_' + arr[i]]) {
                resArr.push(arr[i])
                json['_' + arr[i]] = 1
            }
        } else {
            if (!json['_' + arr[i][att]]) {
                resArr.push(arr[i])
                json['_' + arr[i][att]] = 1
            }
        }
    }
    json = {}
    return resArr
}

// =====================
// 判断两个数组是否一致
// 长度顺序和数值 不适用于多维数组
// 看来要整一个先把多维数组转成简单的数组然后在..进行对比
// =====================
// liuyp 2019年9月20日11:10:17
export const contrast = (arrA, arrB) => {
    if (!Array.isArray(arrA) || !Array.isArray(arrB)) { return false; }
    if (arrA.length !== arrB.length) { return false; }
    for (let i = 0, l = arrA.length; i < l; i++) {
        if (arrA[i] !== arrB[i]) { return false; }
    }
    return true;
}

// =====================
// 多维数组简化为一维数组
// 可搭配使用这个方法用来判断多维数组是否一致
// =====================
// liuyp 2019年9月20日11:10:17
export const arrayFlatten = (ac) => {
    let array = []
    let g = ac
    for (var i = 0; i < g.length; i++) {
        if (g[i] instanceof Array) {
            array = array.concat(arrayFlatten(g[i]))
        } else {
            array.push(g[i])
        }
    }
    return array;
}

// =====================
// 包含确认/交集/差集/并集/补集
// 新数组,不伤源头
// =====================
// liuyp 2018年12月20日11:28:08
// _bs(_.Array_contains([1,2,3,4,5,6],12),result_flag_here
// _bs(_.Array_intersect([1,2,3,4,5,6],[4,5,6,7,8,9]),result_flag_here
// _bs(_.Array_minus([1,2,3,4,5,6],[4,5,6,7,8,9]),result_flag_here
// _bs(_.Array_union([1,2,3,4,5,6],[4,5,6,7,8,9]),result_flag_here
// _bs(_.Array_complement([1,2,3,4,5,6],[4,5,6,7,8,9]),result_flag_here
function arrayFilter (A, fn) {
    fn = fn || function () {}; var arr = [];
    for (var i = 0, l = A.length; i < l; i++) {if (fn(A[i])) {arr.push(A[i])}}
    return arr
}
function arrayContains (A, V) {
    var i = A.length; while (i--) {if (A[i] === V) { return true }} return false
}
export function arrayIntersect (a, b) {
    return arrayFilter(arrayUnique(a), function (o) {return arrayContains(b, o)})
}
export function arrayMinus (a, b) {
    return arrayFilter(arrayUnique(a), function (o) {return !arrayContains(b, o)})
}
export function arrayUnion (a, b) {
    return arrayUnique(a.concat(b))
}
export function arrayComplement (a, b) {
    return arrayMinus(arrayUnion(a, b), arrayIntersect(a, b))
}
// =====================
// 数组洗牌
// =====================
// liuyp 2018年12月20日11:28:08
export function arrayShuffle (o) { // v1.0
    for (var j, x, i = o.length; i; j = Math.floor(Math.random() * i), x = o[--i], o[i] = o[j], o[j] = x) {
        // 666
    }
    return o;
}

// ===================== // ===================== // =====================// ===================== // ===================== // =====================
// ===================== // ===================== // =====================// ===================== // ===================== // =====================
// ===================== // ===================== // =====================// ===================== // ===================== // =====================

// forEach
export const forEach = (arr, fn) => {
    if (!arr.length || !fn) return
    let i = -1
    let len = arr.length
    while (++i < len) {
        let item = arr[i]
        fn(item, i, arr)
    }
}
// 两个数组的并集, 两个数组的元素为数值或字符串
export const getUnion = (arr1, arr2) => {
    return Array.from(new Set([...arr1, ...arr2]))
}
