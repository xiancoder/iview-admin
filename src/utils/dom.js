/**
* @returns {String} 当前浏览器名称
*/
export const getExplorer = () => {
    const ua = window.navigator.userAgent
    const isExplorer = (exp) => {
        return ua.indexOf(exp) > -1
    }
    if (isExplorer('MSIE')) return 'IE'
    else if (isExplorer('Firefox')) return 'Firefox'
    else if (isExplorer('Chrome')) return 'Chrome'
    else if (isExplorer('Opera')) return 'Opera'
    else if (isExplorer('Safari')) return 'Safari'
}
/**
* @description 绑定事件 on(element, event, handler)
*/
export const on = (function () {
    if (document.addEventListener) {
        return function (element, event, handler) {
            if (element && event && handler) {
                element.addEventListener(event, handler, false)
            }
        }
    } else {
        return function (element, event, handler) {
            if (element && event && handler) {
                element.attachEvent('on' + event, handler)
            }
        }
    }
})()
/**
* @description 解绑事件 off(element, event, handler)
*/
export const off = (function () {
    if (document.removeEventListener) {
        return function (element, event, handler) {
            if (element && event) {
                element.removeEventListener(event, handler, false)
            }
        }
    } else {
        return function (element, event, handler) {
            if (element && event) {
                element.detachEvent('on' + event, handler)
            }
        }
    }
})()
/// ///////////////////////////////////////////////////////////////////////////
// 阻止事件冒泡 防止默认行为
// =====================
// liuyp 2018年12月27日19:44:45
/// ///////////////////////////////////////////////////////////////////////////
export const eventStop = (ev) => {
    /* IE和Chrome下是window.event FF下是e */
    ev = ev || window.event;
    /* 防止默认行为 */
    if (ev.preventDefault) { ev.preventDefault() } else { ev.returnValue = false; }
    /* 阻止事件冒泡 */
    if (ev.stopPropagation) { ev.stopPropagation() } else { ev.cancelBubble = true; }
}
/// ///////////////////////////////////////////////////////////////////////////
// 事件发生的目标
// =====================
// liuyp 2018年12月27日19:44:45
/// ///////////////////////////////////////////////////////////////////////////
export const eventTarget = (ev) => {
    ev = ev || window.event;
    /* IE和Chrome下是srcElement FF下是target */
    return ev.target || ev.srcElement;
}

// scrollTop animation
export const scrollTop = (el, from = 0, to, duration = 500, endCallback) => {
    if (!window.requestAnimationFrame) {
        window.requestAnimationFrame = (
            window.webkitRequestAnimationFrame ||
            window.mozRequestAnimationFrame ||
            window.msRequestAnimationFrame ||
            function (callback) {
                return window.setTimeout(callback, 1000 / 60)
            }
        )
    }
    const difference = Math.abs(from - to)
    const step = Math.ceil(difference / duration * 50)

    const scroll = (start, end, step) => {
        if (start === end) {
            endCallback && endCallback()
            return
        }

        let d = (start + step > end) ? end : start + step
        if (start > end) {
            d = (start - step < end) ? end : start - step
        }

        if (el === window) {
            window.scrollTo(d, d)
        } else {
            el.scrollTop = d
        }
        window.requestAnimationFrame(() => scroll(d, end, step))
    }
    scroll(from, to, step)
}

export const findNodeUpperByClasses = (ele, classes) => {
    let parentNode = ele.parentNode
    if (parentNode) {
        let classList = parentNode.classList
        if (classList && classes.every(className => classList.contains(className))) {
            return parentNode
        } else {
            return findNodeUpperByClasses(parentNode, classes)
        }
    }
}

const showTitle = (item, vm) => {
    let { title, __titleIsFunction__ } = item.meta
    if (!title) return
    if (useI18n) {
        if (title.includes('{{') && title.includes('}}') && useI18n) title = title.replace(/({{[\s\S]+?}})/, (m, str) => str.replace(/{{([\s\S]*)}}/, (m, _) => vm.$t(_.trim())))
        else if (__titleIsFunction__) title = item.meta.title
        else title = vm.$t(item.name)
    } else title = (item.meta && item.meta.title) || item.name
    return title
}
