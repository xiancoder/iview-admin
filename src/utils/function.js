// 防抖
export const _debounce = (fn, delay) => {
    delay = delay || 200;
    var timer;
    return function () {
        var th = this;
        var args = arguments;
        if (timer) {
            clearTimeout(timer);
        }
        timer = setTimeout(function () {
            timer = null;
            fn.apply(th, args);
        }, delay);
    };
}
// 节流
export const _throttle = (fn, interval) => {
    var last;
    var timer;
    interval = interval || 200;
    return function () {
        var th = this;
        var args = arguments;
        var now = +new Date();
        if (last && now - last < interval) {
            clearTimeout(timer);
            timer = setTimeout(function () {
                last = now;
                fn.apply(th, args);
            }, interval);
        } else {
            last = now;
            fn.apply(th, args);
        }
    };
}
/**
 * @param {Number} times 回调函数需要执行的次数
 * @param {Function} callback 回调函数
 */
export const doCustomTimes = (times, callback) => {
    let i = -1
    while (++i < times) { callback(i) }
}
