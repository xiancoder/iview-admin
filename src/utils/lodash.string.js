'use strict';
import _ from 'lodash'

export const camelCase = _.camelCase
export const upperFirst = _.upperFirst
export const repeat = _.repeat
