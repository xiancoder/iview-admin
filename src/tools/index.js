import Vue from 'vue'
import { router } from '@/router'
import { Message, Modal, LoadingBar } from 'iview'
import { Store } from '@/store' // 状态管理 -挂载$stroe
import { type } from '@/utils/object'
/// ////////////////////////////////////////////////////
// 常用的操作封装
/// ////////////////////////////////////////////////////
export const goto = (config) => { // 弹框提示
    const current = router.history.current
    if (type(config) === 'string') config = { name: config }
    if (config.name === current.name || config.path === current.path) {
        return router.replace(config).catch(err => {
            if (err.name === 'NavigationDuplicated') return console.log('仙', '路由重复打开小问题')
        })
    }
    router.push(config)
}
export const alert = (msg) => { // 弹框提示
    Message.info({ content: msg || '处理中，请稍后' })
}
export const success = (msg) => { // 成功提示
    Message.success({ content: msg || '保存成功' })
}
export const error = (msg) => { // 错误提示
    Message.error({ content: msg || '保存失败', duration: 10, closable: true })
}
export const confirm = (msg) => { // 二次确认框
    return new Promise((resolve, reject) => {
        Modal.confirm({
            title: '请确认',
            content: '<p>' + msg + '</p>',
            onOk: resolve || function () {},
            onCancel: reject || function () {}
        })
    })
}
export const jumpto = (name, obj) => { // 跳转
    if (obj) { router.push({ name, obj }) } else if (name) { router.push({ name }) } else { window.location.reload() }
}
export const saveParamState = (obj) => { // 保存当前参数
    const time = new Date().getTime()
    const name = router.history.current.name
    const paramList = Store.state.system.paramList
    paramList[name + time] = obj
    const query = { 'search': time }
    router.push({ name, query })
}
export const getParamState = () => { // 获取当前参数
    const time = router.history.current.query.search
    if (time) {
        const name = router.history.current.name
        const paramList = Store.state.system.paramList
        const obj = paramList[name + time]
        return obj
    }
    return {}
}
export const h = { // 通用渲染格式 for 表格 (即将废弃)
    // vue 2.5.14 以上就不再支持h.rander 直接返回字符串了
    // 渲染头像 用户名 电话
    // 参数 头像, 用户名, 电话, 职位
    avatar1: (a, b, c, d) => {
        return (h, params) => {
            const row = params.row;
            return h('div', {}, [
                h('div', {
                    style: { 'display': 'inline-block', 'vertical-align': 'middle', 'margin-right': '10px' }
                }, [
                    row[a]
                        ? h('Avatar', { props: { src: row[a] } })
                        : h('Avatar', { style: { color: '#f56a00', 'background-color': '#fde3cf' } }, row[b][0])
                ]),
                h('div', {
                    style: { 'display': 'inline-block', 'vertical-align': 'middle' }
                }, [
                    h('div', (row[b] || '-') + ' ' + (row[c] || '')),
                    row.department.map(item => {
                        return h('Tag', {
                            props: { color: 'blue' },
                            style: { 'text-align': 'center', 'padding': '1px 4px', 'display': 'inline-block' }
                        }, item[d]);
                    })
                ])
            ])
        }
    },
    // 渲染字符串列表 默认逗号分隔
    // 字符串, 分隔符
    strList1: (a, b) => {
        b = b || ','
        return (h, params) => {
            const row = params.row;
            return h('div', {
                style: { 'display': 'inline-block', 'vertical-align': 'middle' }
            }, [
                row[a] && row[a].split(b).map(item => {
                    return h('Tag', {
                        style: { 'text-align': 'center', 'padding': '1px 4px', 'display': 'inline-block' }
                    }, item);
                })
            ])
        }
    },
    // 正常渲染 默认空值显示 -
    // 值, 默认符
    defaultH: (a, b) => {
        b = b || '-'
        return (h, params) => {
            const row = params.row;
            return h('div', row[a] || b);
        }
    },
    // 读取数字显示对应 没有队形默认显示 -
    // 值, 对应数组, 默认符
    readArr: (a, b, c) => {
        c = c || '-'
        return (h, params) => {
            const row = params.row;
            const text = b[row[a]];
            return h('div', text || c);
        }
    },
    end: null // 错误占位符
}
export const LoadingBarRun = (flag) => { // 顶部进度条执行
    if (flag) {
        LoadingBar.start()
    } else {
        LoadingBar.finish()
    }
}
Vue.prototype.$tool = {
    goto,
    alert,
    success,
    error,
    confirm,
    jumpto,
    saveParamState,
    getParamState,
    h,
    LoadingBarRun
}
