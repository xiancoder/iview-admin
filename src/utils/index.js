'use strict';

import * as array from './base.array.js'
import * as string from './base.string.js'
import * as date from './base.date.js'
import * as date2 from './moment.date.js'
import * as number from './base.number.js'
import * as object from './base.object.js'
import * as dom from './base.dom.js'
import regExp from './base.regexp.js'
import * as event from './base.event.js'
// import * as cookie from './base.cookie.js'
import * as func from './base.function.js'
import exCel from './base.excel.js'

export const unique = array.unique
export const uniqueObj = array.uniqueObj
export const arrayUnique = array.arrayUnique
export const arrayContrast = array.contrast
export const arrayFilter = array.arrayFilter
export const arrayContains = array.arrayContains
export const arrayIntersect = array.arrayIntersect
export const arrayMinus = array.arrayMinus
export const arrayUnion = array.arrayUnion
export const arrayComplement = array.arrayComplement
export const arrayShuffle = array.arrayShuffle

export const dateFormater = date.dateFormater
export const sevenRange = date.sevenRange
export const todayMouth = date.todayMouth
export const dateAdd = date.dateAdd
export const dateSub = date.dateSub
export const date2all = date.date2all
export const timestamp2Date = date.timestamp2Date
export const dateEqual = date.dateEqual

export const dateFormat = date2.dateFormat
export const howManyToToday = date2.howManyToToday
export const computeRange = date2.computeRange
export const computeRangeByMouth = date2.computeRangeByMouth

export const type = object.type
export const extend = object.extend
export const extendF = object.extendF
export const objEqual = object.objEqual
export const deepClone = object.deepClone

export const fillup2Digit = number.fillup2Digit
export const number2Chinese = number.number2Chinese
export const money2Chinese = number.money2Chinese
export const add = number.add
export const sub = number.sub
export const mul = number.mul
export const div = number.div
export const cnNumber = number.cnNumber
export const idCardNumCompute = number.idCardNumCompute
export const estimate = number.estimate
export const compare = number.compare
export const thousand = number.thousand
export const bankCartNum = number.bankCartNum
export const toDecimalForce = number.toDecimalForce
export const toDecimalFormat = number.toDecimalFormat

export const getUrlParam = dom.getUrlParam
export const scrollTop = dom.scrollTop
export const print = dom.print
export const findNodeUpperByClasses = dom.findNodeUpperByClasses

export const debounce = func.debounce
export const throttle = func.throttle
export const nothing = func.nothing

export const regexp = regExp
console.log(regexp)

export const excel = exCel

export const ramdomString = string.ramdomString
export const url2obj = string.url2obj
export const obj2url = string.obj2url
export const encodeBase64 = string.encodeBase64
export const decodeBase64 = string.decodeBase64
export const capitalize = string.capitalize

export const on = event.on
export const off = event.off
export const eventStop = event.eventStop
