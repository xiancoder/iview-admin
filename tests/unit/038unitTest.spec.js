/**
* 功能：单元测试页面
* 作者：liuyp
* 时间：2019年10月10日14:56:59
* 语句: npm run test:unit tests/unit/038unitTest.spec.js
*/
import { mount, createLocalVue } from '@vue/test-utils'
import fromTest from '@V/exp2/038unitTest'
import iView from 'iview'
import '@/tools' // 常用工具 -挂载$tool
import '@/utils' // 常用方法 -挂载$util
// mock掉整个axios模块
// 返回值在使用的时候自定义
jest.mock('@/api/unit.js', () => ({
    getDetail: jest.fn(() => Promise.resolve({
        input1: '1111', // 输入框测试
        input2: '1111' // 输入框测试
    })),
    submit: jest.fn(() => Promise.resolve({
        res: 1
    }))
}))
// 创建临时Vue实例，挂载组件中使用的插件
const localVue = createLocalVue()
localVue.use(iView)
// skip语法可以跳过测试，不用大规模注释代码
describe.skip('skip', () => {
    let foo = false
    it('slip是否执行', () => {
        expect(foo).toBe(false)
    })
});
describe('[单元测试038.表单单元流程]', () => {
    let wrapper
    beforeEach(() => {
        wrapper = mount(fromTest, { localVue })
    })
    afterEach(() => {
        wrapper.destroy()
    })
    // 倒着一步一步赋值验证
    it('提交ajax成功后应该触发跳转方法', () => {
        const mockFn1 = jest.fn()
        wrapper.setMethods({ '$tool': { 'jumpto': mockFn1 } })
        wrapper.vm.submit()
        wrapper.vm.$nextTick(() => { // 异步监听
            expect(wrapper.vm.testResult2).toEqual([4]) // 输入框测试
            expect(mockFn1).toBeCalled() // 断言函数被触发且只触发一次
        })
    })
    it('验证成功后提交ajax(还没写)', () => {
        expect(1 + 1).toEqual(2)
    })
    it('点击按钮触发验证', () => {
        return wrapper.vm.init().then(result => {
            const mockFn2 = jest.fn()
            wrapper.setMethods({ 'submit': mockFn2 })
            const buttonValidate = wrapper.find('.flagBtn')
            buttonValidate.trigger('click')
            wrapper.vm.$nextTick(() => { // 异步监听
                expect(wrapper.vm.testResult).toEqual([1, 2, 3]) // 输入框测试
                expect(mockFn2).toBeCalled() // 断言函数被触发且只触发一次
            })
        })
    })
    it('iview输入框可以根据事件修改内容触发监听', () => {
        const wrapperArray = wrapper.findAll('.flag')
        wrapperArray.wrappers.forEach(wrapper => {
            wrapper.vm.$emit('input', '2222')
        })
        expect(wrapper.vm.frm).toEqual({ input1: '2222', input2: '2222' }) // 输入框测试
    })
    it('表单初始化可以给变量赋初始值', () => {
        wrapper.vm.init().then(result => {
            expect(wrapper.vm.frm).toEqual({ input1: '1111', input2: '1111' }) // 输入框测试
        })
    })
    it('验证保存快照', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })
})
