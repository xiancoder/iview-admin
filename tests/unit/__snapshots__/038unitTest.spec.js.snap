// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[单元测试038.表单单元流程] 验证保存快照 1`] = `
<div class="blogCss xianFrom">
  <div class="blog">
    <div class="blogTitle">表单的单元测试 (mock赋值 - 模拟操作 - mock提交 - 校验)</div>
    <div class="blogContent">
      <div class="xian-title-tabs">
        <h1>表单名称</h1>
        <div class="ivu-divider ivu-divider-horizontal ivu-divider-default">
          <!---->
        </div>
      </div>
      <form autocomplete="off" class="ivu-form ivu-form-label-right">
        <div class="ivu-form-item"><label class="ivu-form-item-label" style="width: 150px;">正常文本</label>
          <div class="ivu-form-item-content" style="margin-left: 150px;">
            <p></p>
            <!---->
          </div>
        </div>
        <div class="ivu-form-item ivu-form-item-required ivu-form-item-error"><label class="ivu-form-item-label" style="width: 150px;">输入框</label>
          <div class="ivu-form-item-content" style="margin-left: 150px;">
            <div class="flag ivu-input-wrapper ivu-input-wrapper-default ivu-input-type" style="width: 450px;">
              <!---->
              <!----> <i class="ivu-icon ivu-icon-ios-loading ivu-load-loop ivu-input-icon ivu-input-icon-validate" name="fade"></i> <input autocomplete="off" spellcheck="false" type="text" placeholder="请输入XXXX" class="ivu-input ivu-input-default">
              <!---->
            </div>
            <div class="ivu-form-item-error-tip" name="fade">输入框不能为空</div>
          </div>
        </div>
        <div class="ivu-form-item ivu-form-item-required ivu-form-item-error"><label class="ivu-form-item-label" style="width: 150px;">输入框</label>
          <div class="ivu-form-item-content" style="margin-left: 150px;">
            <div class="flag ivu-input-wrapper ivu-input-wrapper-default ivu-input-type" style="width: 450px;">
              <!---->
              <!----> <i class="ivu-icon ivu-icon-ios-loading ivu-load-loop ivu-input-icon ivu-input-icon-validate" name="fade"></i> <input autocomplete="off" spellcheck="false" type="text" placeholder="请输入XXXX" class="ivu-input ivu-input-default">
              <!---->
            </div>
            <div class="ivu-form-item-error-tip" name="fade">输入框不能为空</div>
          </div>
        </div>
        <div class="ivu-form-item"><label class="ivu-form-item-label" style="width: 150px;">测试</label>
          <div class="ivu-form-item-content" style="margin-left: 150px;"><button type="button" class="flagBtn ivu-btn ivu-btn-primary">
              <!---->
              <!----> <span>测试</span></button>
            <!---->
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
